File: tools/poyo/internal/scaffold/scaffold.go
package scaffold

import (
	"fmt"
	"os"
	"path/filepath"
	"poyo-cli/internal/config"
	"poyo-cli/internal/routes"
)

type ScaffoldOptions struct {
	NoView bool
}

type ControllerInfo struct {
	Name   string
	Action string
}

func ScaffoldRouteFiles(name string, files routes.Files, options ScaffoldOptions, controller *ControllerInfo) error {
	pageFullPath := filepath.Join(config.ClientDir, files.React)
	viewFullPath := filepath.Join(config.ServerDir, files.View)

	// 1. React Page
	if _, err := os.Stat(pageFullPath); os.IsNotExist(err) {
		if err := os.MkdirAll(filepath.Dir(pageFullPath), 0755); err != nil {
			return err
		}
		if err := os.WriteFile(pageFullPath, []byte(ReactPage(name)), 0644); err != nil {
			return err
		}
		fmt.Printf("[CREATED] React Page: %s\n", files.React)
	} else {
		fmt.Printf("[EXISTS] React Page: %s\n", files.React)
	}

	// 2. MVC View
	if !options.NoView {
		if _, err := os.Stat(viewFullPath); os.IsNotExist(err) {
			if err := os.MkdirAll(filepath.Dir(viewFullPath), 0755); err != nil {
				return err
			}
			if err := os.WriteFile(viewFullPath, []byte(MVCView(name)), 0644); err != nil {
				return err
			}
			fmt.Printf("[CREATED] MVC View: %s\n", files.View)
		} else {
			fmt.Printf("[EXISTS] MVC View: %s\n", files.View)
		}
	} else {
		fmt.Println("[SKIP] MVC View generation skipped (--no-view)")
	}

	// 3. Controller Injection
	if controller != nil {
		_, err := EnsureController(
			config.ControllersDir,
			controller.Name,
			controller.Action,
			files.View,
		)
		if err != nil {
			// If action exists, we just log it, not fail everything
			if err.Error() == "action already exists" {
				fmt.Printf("[INFO] Action '%s' already exists in %sController\n", controller.Action, controller.Name)
			} else {
				return err
			}
		} else {
			fmt.Printf("[UPDATED] Controller: %sController.cs (Injected action '%s')\n", controller.Name, controller.Action)
		}
	}

	return nil
}

File: packages/poyo/src/scaffold.ts
import fs from "node:fs";
import path from "node:path";
import type { ProjectPaths } from "./config.js";
import type { RouteFile } from "./routes.js";
import {
	actionTemplate,
	controllerTemplate,
	pageTemplate,
	viewTemplate,
} from "./templates.js";

export function findFiles(
	dir: string,
	predicate: (filePath: string) => boolean,
	fileList: string[] = [],
	rootDir = dir,
): string[] {
	if (!fs.existsSync(dir)) return fileList;
	const entries = fs.readdirSync(dir);
	for (const entry of entries) {
		const entryPath = path.join(dir, entry);
		const stat = fs.statSync(entryPath);
		if (stat.isDirectory()) {
			findFiles(entryPath, predicate, fileList, rootDir);
		} else if (predicate(entryPath)) {
			fileList.push(path.relative(rootDir, entryPath).replaceAll("\\", "/"));
		}
	}
	return fileList;
}

export function deleteEmptyParents(filePath: string, rootDir: string): void {
	let dir = path.dirname(filePath);
	while (dir !== rootDir && path.relative(rootDir, dir).length > 0) {
		if (fs.existsSync(dir) && fs.readdirSync(dir).length === 0) {
			fs.rmdirSync(dir);
			dir = path.dirname(dir);
		} else {
			break;
		}
	}
}

export function ensureControllerAction(
	paths: ProjectPaths,
	controllerName: string,
	actionName: string,
	viewPath: string,
): string {
	let safeControllerName = controllerName;
	if (!safeControllerName.endsWith("Controller")) {
		safeControllerName += "Controller";
	}

	const controllerPath = path.join(
		paths.controllersDir,
		`${safeControllerName}.cs`,
	);

	if (!fs.existsSync(controllerPath)) {
		fs.mkdirSync(paths.controllersDir, { recursive: true });
		fs.writeFileSync(
			controllerPath,
			controllerTemplate(safeControllerName, actionName, viewPath),
		);
		return safeControllerName;
	}

	const content = fs.readFileSync(controllerPath, "utf-8");
	const actionRegex = new RegExp(
		`public\\s+(async\\s+Task<)?IActionResult(>)?\\s+${actionName}\\s*\\(`,
		"i",
	);
	if (actionRegex.test(content)) {
		throw new Error(
			`Action '${actionName}' already exists in ${safeControllerName}.cs`,
		);
	}

	const lastBraceIndex = content.lastIndexOf("}");
	if (lastBraceIndex === -1) {
		throw new Error(
			`Could not parse class structure in ${safeControllerName}.cs`,
		);
	}

	const newContent =
		content.slice(0, lastBraceIndex) +
		actionTemplate(actionName, viewPath) +
		content.slice(lastBraceIndex);
	fs.writeFileSync(controllerPath, newContent);
	return safeControllerName;
}

export function scaffoldRouteFiles(
	paths: ProjectPaths,
	name: string,
	files: RouteFile,
	options: { noView?: boolean; controller?: string; action?: string },
): void {
	const pageFullPath = path.join(paths.clientDir, files.react);
	const viewFullPath = path.join(paths.serverDir, files.view);

	if (!fs.existsSync(pageFullPath)) {
		fs.mkdirSync(path.dirname(pageFullPath), { recursive: true });
		fs.writeFileSync(pageFullPath, pageTemplate(name));
	}

	if (options.noView) {
		return;
	}

	if (!fs.existsSync(viewFullPath)) {
		fs.mkdirSync(path.dirname(viewFullPath), { recursive: true });
		fs.writeFileSync(viewFullPath, viewTemplate(name));
	}

	if (options.controller && options.action) {
		ensureControllerAction(
			paths,
			options.controller,
			options.action,
			files.view,
		);
	}
}
